src/main/scala/tsiHost/TsiHostWidgetUnitTest.scala

```diff
@@ -0,0 +1,172 @@
+ * Currently only tests read TSI requests since TLROM can be setup easily with data
+ * that way.
+ */
+class TSIHostWidgetTest(implicit p: Parameters) extends LazyModule {
+  // params matching the configuration for the TSIHost widget
+  val beatBytes = 8
+  val lineBytes = p(PeripheryTSIHostKey).tlLineBytes
+  val serWidth = p(PeripheryTSIHostKey).serialIfWidth
+  val addrSet = p(PeripheryTSIHostKey).tlAddressSet
+  val regionTyp = p(PeripheryTSIHostKey).tlRegionType
+  val numXacts = p(PeripheryTSIHostKey).tlNumTransactions
+
+  // TSIHost widget in host-land connecting to the target Serdes
+  val hostTSIHostWidget = LazyModule(new TLTSIHostWidget(beatBytes))
+
+  // lives in target-land (connects to the TSIHost widget in host-land)
+  val targetSerdes = LazyModule(new TLSerdesser(
+    w = serWidth,
+    clientParams = TLClientParameters(
+      name = "tl-desser",
+      sourceId = IdRange(0, numXacts)),
+    managerParams = TLManagerParameters(
+      address = Seq(addrSet),
+      regionType = regionTyp,
+      supportsGet = TransferSizes(1, lineBytes),
+      supportsPutFull = TransferSizes(1, lineBytes),
+      supportsPutPartial = TransferSizes(1, lineBytes)),
+    beatBytes = beatBytes))
+
+  // rom living in target-land. connects to the client of the serdes
+  val targetRomData = Seq(
+    BigInt("7766554433221100", 16),
+    BigInt("FFEEDDCCBBAA9988", 16),
+    BigInt("0123456789ABCDEF", 16),
+    BigInt("FEDCBA9876543210", 16),
+    BigInt("76543210FDECBA98", 16))
+  val targetRom = LazyModule(new TLROM(0, 64,
+    targetRomData.flatMap(
+      data => (0 until 8).map(i => ((data >> (i * 8)) & 0xff).toByte)),
+    beatBytes = beatBytes))
+
+  // TLFragmenter splits up the multi beat packet into single beats that the left hand side can take
+  // (of size beatBytes).
+  targetRom.node := TLFragmenter(beatBytes, lineBytes) := TLBuffer() := targetSerdes.clientNode
+
+  // fuzzer that lives in target-land that sends random requests to the ram on the host
+  val targetFuzzer = LazyModule(new TLFuzzer(
+    nOperations = 32,
+    inFlight = numXacts))
+
+  // ram living in host-land. for the fuzzer to read and write from
+  val hostRam = LazyModule(new TLTestRAM(
+    address = addrSet,
+    beatBytes = beatBytes))
+
+  // connect fuzzer to the target serdes
+  targetSerdes.managerNode := TLBuffer() := targetFuzzer.node
+  // connect the host node to the host rom
+  hostRam.node := TLFragmenter(beatBytes, lineBytes) := TLBuffer() := hostTSIHostWidget.externalClientNode
+
+  // implementation of the module
+  lazy val module = new LazyModuleImp(this) {
+    // i/o to connect to the unit test interface
+    val io = IO(new Bundle with UnitTestIO)
+
+    // TSI cmd of the form (cmd, lowAddr, highAddr, lowLen, highLen)
+    // note that len is actually the len - 1 (e.g. len of 4 is really 5)
+    // and len is in increments of 32b
+    // O is read, 1 is write
+    val inputTsiSeq = Seq(Seq(0,  0x0, 0, 0x0, 0),
```

Comment:
We probably want a helper object here to generate commands that are of some TSICommand class type. Not exactly this, but something like:

`val inputTsiSeq = Seq(TSIWrite(0x80000000, 0xdeadbeef), TSIRead(0x80000000), TSIWrite(0x80000008, 0xffff0000))`

It may even be worth making your own `TSIFuzzer`

Reply:
Other than this widget will there be a need to create a TSIFuzzer? For now, I think just a helper factory would be good.

Reply:
Maybe not, let's see what it looks like after you make some changes.